STARTTLS: add support for xmpp-server

horazont opened the pull request:
XMPP client-to-server and server-to-server links historically use
different XML namespaces. Some server implementations are strict
about this and will not proceed with the connection attempt when
the client namespace (`jabber:client`) is used on a
server-to-server link.

openssl s_client also supports `xmpp-server`.

horazont committed:
STARTTLS: add support for xmpp-server

XMPP client-to-server and server-to-server links historically use
different XML namespaces. Some server implementations are strict
about this and will not proceed with the connection attempt when
the client namespace (`jabber:client`) is used on a
server-to-server link.

openssl s_client also supports `xmpp-server`.

horazont commented:
As for a test case:

Before:

```
$ ./testssl.sh --openssl=/usr/bin/openssl --starttls xmpp --xmpphost conference.jabber.org hermes2.jabber.org:5269

No engine or GOST support via engine with your /usr/bin/openssl

###########################################################
    testssl.sh       3.0.1 from https://testssl.sh/
    (af86cce 2020-04-15 12:35:51 -- )

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.1.1f  31 Mar 2020" [~79 ciphers]
 on sinistra:/usr/bin/openssl
 (built: "Mar 31 21:59:59 2020", platform: "debian-amd64")


 Start 2020-04-19 15:10:14        -->> 208.68.163.218:5269 (hermes2.jabber.org) <<--

 Further IP addresses:   2605:da00:5222:5269::2:2
 rDNS (208.68.163.218):  xmpp.org.

 hermes2.jabber.org:5269 doesn't seem to be a TLS/SSL enabled server
 The results might look ok but they could be nonsense. Really proceed ? ("yes" to continue) -->
```

After:
```
$ ./testssl.sh --openssl=/usr/bin/openssl --starttls xmpp-server --xmpphost conference.jabber.org hermes2.jabber.org:5269

No engine or GOST support via engine with your /usr/bin/openssl

###########################################################
    testssl.sh       3.0.1 from https://testssl.sh/
    (76fd21d 2020-04-19 15:08:13 -- )

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.1.1f  31 Mar 2020" [~79 ciphers]
 on sinistra:/usr/bin/openssl
 (built: "Mar 31 21:59:59 2020", platform: "debian-amd64")


 Start 2020-04-19 15:11:21        -->> 208.68.163.218:5269 (hermes2.jabber.org) <<--

 Further IP addresses:   2605:da00:5222:5269::2:2
 rDNS (208.68.163.218):  xmpp.org.
 Service set:            STARTTLS via XMPP-SERVER (XMPP domain=\'conference.jabber.org\')

 Testing protocols via sockets

 SSLv2      not offered (OK)
 SSLv3      not offered (OK)
 TLS 1      ^C

```

Note that I’m using my system-installed `openssl` because the one in the repository does not support `xmpp-server`. I don’t know how to fix that.

drwetter commented:
Hi @horazont ,

thanks! 

Wasn't aware of this function, never looked into it. So I thought ``xmpp`` would do it also for s2s. 

The problem with the distributed binary is that it would need to be patched, see e.g. https://github.com/openssl/openssl/issues/3980 -- or what the current version of the code is. See https://github.com/drwetter/testssl.sh/tree/3.1dev/bin/ . But it hasn't been determined whether we'll continue using a patched 1.0.2 version or patch a newer version which comes e.g. with TLS 1.3.

To get your PR merged perfectly, would you mind:

1) amend the PR in ``help()``
2) amend the PR with documentation in ``doc/testssl.1.md``

Then: It's a minor addition but not sure yet I want this new feature in the stable 3.0 branch. In any case it needs to be in 3.1dev.

horazont commented:
@drwetter 
So, the action plan for me is:

1. Rebase on 3.1dev
2. Amend the help() function so that it mentions xmpp-server
3. Amend the manpage so that it also mentions xmpp-server.

Correct? Other than that, the PR is good?

drwetter commented on testssl.sh:
```diff
@@ -10349,8 +10352,8 @@ fd_socket() {
-               xmpp|xmpps) # XMPP, see https://tools.ietf.org/html/rfc6120
-                    starttls_xmpp_dialog
+               xmpp|xmpps|xmpp-server) # XMPP, see https://tools.ietf.org/html/rfc6120
+                    starttls_xmpp_dialog "$STARTTLS_PROTOCOL"
```
Minor thing: ``starttls_xmpp_dialog()`` isn't requesting an argument and even the global ``$STARTTLS_PROTOCOL`` isn't needed.

drwetter commented:
5x yes -- I think so. 

Probably the project also could need a Travis/Ci test (``t/25_baseline_starttls.t``) but this is optional and would atm only work for sockets, right?

drwetter commented:
If I was asking too much this is also fine, then I would just do it. Let me know.

horazont commented:
@drwetter Uh, sorry, no. I just got distracted and this scrolled waaaay out of the 30 line window of my inbox I look at :D. I’ll get right at it.

horazont commented:
@drwetter I rebased and amended the docs and did some smoketesting to see that it still does the right thing. I also changed the target of the PR.

I wasn’t able to quickly dig up since when openssl support xmpp-server. If you happen to know, I’ll gladly add that to the manpage source.

drwetter commented:
Thank you! 

> I wasn’t able to quickly dig up since when openssl support xmpp-server. If you happen to know, I’ll gladly add that to the manpage source.

That might be interesting for us to find out (first choice is their [changelog](https://www.openssl.org/news/changelog.html) , then github). For the m an page of testslssl.sh it's not important. Similar to HAS_XMPP I'll try to add a variable though like HAS_XMPP_SERVER which tests whether the binary supports that. We'll rather check for the support, not version

drwetter commented:
FYI This is how it will look like when there's no support for xmpp-server (pending PR #1601 with HAS_XMPP_SERVER)

![Screenshot_20200501_184651](https://user-images.githubusercontent.com/8036727/80823254-9a89a780-8bdc-11ea-96c6-527b836ef860.png)

When xmpp-server made it into openssl seems in the first glance a bit awkward. git blame / git log list July 2015(!) but it's not in 1.0.2 -- which was released in the beginning of 2015

drwetter commented:
FYI: https://github.com/openssl/openssl/issues/3980#issuecomment-316804142 I didn't include this one yet

drwetter commented:
I added in https://github.com/drwetter/openssl-1.0.2.bad https://github.com/openssl/openssl/issues/3980#issuecomment-316804142 but not n the binary yet.

I would love to provide new binaries soon and a a few further steps are depended on that. So if you @horazont can sacrifice a few minutes for testing that would be much appreciated.

horazont commented:
@drwetter I'll see if I can give it a shot on xmpp.net/preview, [which is based on testssl.sh](https://github.com/horazont/testxmpp/), over the weekend, thanks a lot!

drwetter commented:
Perfect. 

FYI: likely / hopefully this will be the last binary release as binaries are a hassle. So that's why I though a thorough look from you makes sense ;-)

horazont commented:
Hm, am I holding it wrong?

```
I have no name!@66b3ec98c279:/tmp$ wget https://testssl.sh/openssl.Linux.x86_64.xmppserver
--2022-08-27 08:18:09--  https://testssl.sh/openssl.Linux.x86_64.xmppserver
Resolving testssl.sh (testssl.sh)... 81.169.166.184, 2a01:238:4308:a920:1000:0:e571:51
Connecting to testssl.sh (testssl.sh)|81.169.166.184|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4459639 (4.3M) [application/octet-stream]
Saving to: 'openssl.Linux.x86_64.xmppserver'

openssl.Linux.x86_64.xmpp 100%[=====================================>]   4.25M  3.96MB/s    in 1.1s

2022-08-27 08:18:11 (3.96 MB/s) - 'openssl.Linux.x86_64.xmppserver' saved [4459639/4459639]

I have no name!@66b3ec98c279:/tmp$ chmod +x openssl.Linux.x86_64.xmppserver
I have no name!@66b3ec98c279:/tmp$ ./openssl.Linux.x86_64.xmppserver s_client
Error configuring OpenSSL
30435392:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(libssl_conf.so): libssl_conf.so: cannot open shared object file: No such file or directory
30435392:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
30435392:error:0E07506E:configuration file routines:MODULE_LOAD_DSO:error loading dso:conf_mod.c:271:module=ssl_conf, path=ssl_conf
30435392:error:0E076071:configuration file routines:MODULE_RUN:unknown module name:conf_mod.c:212:module=ssl_conf
I have no name!@66b3ec98c279:/tmp$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

drwetter commented:
It ran on a newer system of mine and it should be statically compiled on a legacy VM .

Can check this evening earliest unfortunately. 

Am 27. August 2022 10:18:58 MESZ schrieb "Jonas Schäfer" ***@***.***>:
>Hm, am I holding it wrong?
>
>```
>I have no ***@***.***:/tmp$ wget https://testssl.sh/openssl.Linux.x86_64.xmppserver
>--2022-08-27 08:18:09--  https://testssl.sh/openssl.Linux.x86_64.xmppserver
>Resolving testssl.sh (testssl.sh)... 81.169.166.184, 2a01:238:4308:a920:1000:0:e571:51
>Connecting to testssl.sh (testssl.sh)|81.169.166.184|:443... connected.
>HTTP request sent, awaiting response... 200 OK
>Length: 4459639 (4.3M) [application/octet-stream]
>Saving to: 'openssl.Linux.x86_64.xmppserver'
>
>openssl.Linux.x86_64.xmpp 100%[=====================================>]   4.25M  3.96MB/s    in 1.1s
>
>2022-08-27 08:18:11 (3.96 MB/s) - 'openssl.Linux.x86_64.xmppserver' saved [4459639/4459639]
>
>I have no ***@***.***:/tmp$ chmod +x openssl.Linux.x86_64.xmppserver
>I have no ***@***.***:/tmp$ ./openssl.Linux.x86_64.xmppserver s_client
>Error configuring OpenSSL
>30435392:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(libssl_conf.so): libssl_conf.so: cannot open shared object file: No such file or directory
>30435392:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
>30435392:error:0E07506E:configuration file routines:MODULE_LOAD_DSO:error loading dso:conf_mod.c:271:module=ssl_conf, path=ssl_conf
>30435392:error:0E076071:configuration file routines:MODULE_RUN:unknown module name:conf_mod.c:212:module=ssl_conf
>I have no ***@***.***:/tmp$ cat /etc/os-release
>PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
>NAME="Debian GNU/Linux"
>VERSION_ID="11"
>VERSION="11 (bullseye)"
>VERSION_CODENAME=bullseye
>ID=debian
>HOME_URL="https://www.debian.org/"
>SUPPORT_URL="https://www.debian.org/support"
>BUG_REPORT_URL="https://bugs.debian.org/"
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/drwetter/testssl.sh/pull/1575#issuecomment-1229149093
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

drwetter commented:
Are you trying it within a container?

horazont commented:
Yep, the dockerfile is here: https://github.com/horazont/testxmpp/blob/master/docker/testssl.Dockerfile

It should be mostly standard debian bullseye.

drwetter commented:
I honestly don't know what the problem is on your side

![image](https://user-images.githubusercontent.com/8036727/187088285-fa3e9e7d-7099-469c-9629-22ee1bedcf4e.png)

![image](https://user-images.githubusercontent.com/8036727/187088476-35280465-f626-4d20-9beb-dcc261d1802c.png)

horazont commented:
Thanks for looking into it, I'll give it another shot over the course of next week!

drwetter commented:
PS:

prompt> ldd openssl.Linux.x86_64.xmppserver
	not a dynamic executable
prompt> file openssl.Linux.x86_64.xmppserver
openssl.Linux.x86_64.xmppserver: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, for GNU/Linux 2.6.24, BuildID[sha1]=ee43229ab83d97f4791e7777d4fca7b3baa23f3d, not stripped
prompt>
```

horazont commented:
@drwetter `file` and `ldd` say the same for me, and I can also run the openssl command itself and also `version -a`, but `s_client` tries to load some library apparently:

```
I have no name!@80e6b45d5742:/$ sha256sum /tmp/openssl
26f480d645c2e17b7fde0b3412c7cc1af6960194caa75260fe0c4a8f3b02dc1e  /tmp/openssl
I have no name!@80e6b45d5742:/$ /tmp/openssl version -a
OpenSSL 1.0.2-chacha (1.0.2k-dev)
built on: Tue Aug 23 13:45:06 2022
platform: linux-x86_64
options:  bn(64,64) md2(int) rc4(16x,int) des(idx,cisc,16,int) idea(int) blowfish(idx)
compiler: gcc -I. -I.. -I../include  -DZLIB -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_USE_BUILD_DATE -DOPENSSL_USE_IPV6 -static -Wa,--noexecstack -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_EXPERIMENTAL_JPAKE -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
I have no name!@80e6b45d5742:/$ /tmp/openssl s_client
Error configuring OpenSSL
32704576:error:25066067:DSO support routines:DLFCN_LOAD:could not load the shared library:dso_dlfcn.c:187:filename(libssl_conf.so): libssl_conf.so: cannot open shared object file: No such file or directory
32704576:error:25070067:DSO support routines:DSO_load:could not load the shared library:dso_lib.c:233:
32704576:error:0E07506E:configuration file routines:MODULE_LOAD_DSO:error loading dso:conf_mod.c:271:module=ssl_conf, path=ssl_conf
32704576:error:0E076071:configuration file routines:MODULE_RUN:unknown module name:conf_mod.c:212:module=ssl_conf
I have no name!@80e6b45d5742:/$ ldd /tmp/openssl
        not a dynamic executable
```

This also happens on my debian/testing (bookworm) host system, outside the container.

drwetter commented:
Try e.g.

OPENSSL_CONF='' ./openssl.Linux.x86_64.xmppserver s_client -connect testssl.sh:443

horazont commented:
A-ha! That does seem to work standalone (e.g. `/tmp/openssl s_client -starttls xmpp-server -xmpphost zombofant.net -connect xmpp-lysithea.sotecware.net:5270`), but:

1. The detection of xmpp-server support in `     grep -q 'xmpp-server' $s_client_starttls_has && HAS_XMPP_SERVER=true` breaks because `xmpp-server` is not printed in the output.
2. I'm not so sure if this will then work in any case, given that testssl.sh would like to pass a non-empty `$OPENSSL_CONF` to enable GOST?

drwetter commented:
1.: I'll look into it tomorrow
2.: that was just an example to avoid that Debian uses its own incompatible openssl.cnf . In testssl.sh we provide a non-empty config file, so the the new binary can be used as a 1:1  replacement

drwetter commented:
Looks to me later versions are somehow different in their response I used for detection. The PR above should fix the detection for the legacy version 1.0.2. Could you test that please?

horazont commented:
Looks great in combination with #2222! Thanks a lot for the backport!

drwetter commented:
Cool! Thank you.

Will work on the binaries then during the course of the next days.

FYI: On a Mac LibreSSL 2.8.3 doesn't support this option. And on a Silicon CPU it's not feasible to compile openssl-1.0.2-bad. The architecture is not supported.
